### DISCUSSION TITLE
Topic modelling on a mixture of short and long documents?

### DISCUSSION THREAD

**Discussion by francesca-lucas** [MaartenGr/BERTopic#1093] [Q&A]
Hi,

First of all, I'm very happy with BERTopic, how easily it is to get started and then how modular it is and all the possibilities this gives. :) 

I'm also very happy with this discussion board, I don't have many people in my circle working on topic modelling so I'm curious to hear your input and ideas! (And apologies for the wall of text, I tried to be as complete as possible...)

**Introduction**:
I would like to extract the topics from a dataset consisting of news articles and political documents in Dutch (typically 20k - 50k documents). 

This means we have a combination of short and long documents/texts. Some have the length of one paragraph, whereas other docs are multiple pages long. The long documents can contain either one topic or a mixture of topics. (Technically, all documents could contain a mixture of topics, and this obviously also depends on the topic granularity, but I made the assumption that most shorter docs/news articles contain one topic.)

News articles vs political documents contain different ways of phrasing. I suppose this will be less of a problem due to the semantic embeddings (instead of bag of words approaches), but this too could complicate things.

Suppose we would use the whole documents directly to train the model. This gives a few problems. The longer documents probably contain a mixtures of topics, which is conflicting with the BERTopic assumption of one topics per doc. Furthermore, the longer docs could get truncated due to the token limit (although this token limit depends on the chosen embedding), which means we could miss parts of the document and therefore topics. 

One direction could be to split the dataset into two: the short docs (news articles) and the long docs (political documents) and train two separate topic models. This doesn't solve all problems but could potentially alleviate some of the differences in doc length and type of language used. However, this would mean having to match the topics from both models afterwards manually. This is less than ideal.

**Challenge**: how to train a model on a mixture of short and long documents without truncating or overrepresenting the long documents. 

Some (naive) ideas/approaches, using the combined dataset: 

1. Sentencize each document. Use all the sentences as training data. Problem: longer documents get overrepresented in both the creation of topics as well as the topic frequency counts.

2. Sentencize each document. Then, take a fixed size (random) sample of sentences from all documents, e.g. 10 sentences for each doc. Problem: high risk of missing topics in the longer documents. 

3. Sentencize each document. Use all the sentences as training data. Then, after training, approximate the topic distribution for each document using `topic_model.approximate_distribution(docs)`. Problem: longer documents get overrepresented only in the creation of topics.

So far I thought option 3 seems like an interesting option, but I made a test run for this and unfortunately it gave rather bad results. With 'bad results' I mean that the approximate topic distribution seems to either predict lower-frequency topics, or topics that are "polluted" or meaningless. This could also be an indication that the topic model itself needs some work first. 
Side note: from a data science point of view it seems bad practice to train the model on sentences and then predict on whole documents, but due to the sliding window nature of `.approximate_distribution()` I thought it would be ok. 

**Question**:
I was wondering if you have any feedback on these ideas, if I made an error in my reasoning somewhere, or if you have any other ideas? 
Did you ever combine short and long documents, and if so, how did you do that?

**Comment by MaartenGr** [comment]
> I would like to extract the topics from a dataset consisting of news articles and political documents in Dutch (typically 20k - 50k documents).

Great, a Dutch dataset! I have not seen that many before aside from my own analyses. 

> One direction could be to split the dataset into two: the short docs (news articles) and the long docs (political documents) and train two separate topic models. This doesn't solve all problems but could potentially alleviate some of the differences in doc length and type of language used. However, this would mean having to match the topics from both models afterwards manually. This is less than ideal.

I think that you nicely identified another complexity rather than documents of different lengths and that is that you essentially have two datasets with one being generally longer than the other. Do note though that there might not be a need to match topics manually afterward as you can find overlapping topics with the trick described [here](https://maartengr.github.io/BERTopic/getting_started/tips_and_tricks/tips_and_tricks.html#finding-similar-topics-between-models). 

> Challenge: how to train a model on a mixture of short and long documents without truncating or overrepresenting the long documents.

Before going into the ideas/approaches that you have is that it is important to understand what is means to "overrepresent" something in the context of BERTopic, and more specifically, in the context of clustering. Depending on the clustering algorithm that you use, it might not be a problem to have different distributions of clusters as they are still modeled anyway. For example, in HDBSCAN, it matters to a lesser extent that the distribution of clusters is different. You actually see that quite often with HDBSCAN as it tends to capture clusters of many different sizes. So the "overrepresentation" in the clustering task might not actually be that big of a problem. The "overrepresentation" with respect to the document frequency is indeed a different story but I'll come back to that a bit later. 

> Sentencize each document. Use all the sentences as training data. Problem: longer documents get overrepresented in both the creation of topics as well as the topic frequency counts.

As mentioned above, I think the main issue would here be the topic frequency counts. Some topics might be overestimated as they appear multiple times in documents. To counteract this, I would advise aggregating these counts outside of BERTopic and potentially then update the topic frequencies directly in BERTopic. How you do the aggregation is then up to you. You can simply count the number of documents that contain the topic at least once or even take the most frequent topic in a document. 

> Sentencize each document. Then, take a fixed size (random) sample of sentences from all documents, e.g. 10 sentences for each doc. Problem: high risk of missing topics in the longer documents.

Although this might not be necessary depending on the clustering algorithm, it might work out if the long documents have very similar topics and if you are not interested in micro-topics. Larger topics will be typically kept with this I believe. 

> Sentencize each document. Use all the sentences as training data. Then, after training, approximate the topic distribution for each document using topic_model.approximate_distribution(docs). Problem: longer documents get overrepresented only in the creation of topics.
> So far I thought option 3 seems like an interesting option, but I made a test run for this and unfortunately it gave rather bad results. With 'bad results' I mean that the approximate topic distribution seems to either predict lower-frequency topics, or topics that are "polluted" or meaningless. This could also be an indication that the topic model itself needs some work first.

It might be worthwhile to play around with the `min_similarity` parameter of `.approximate_distribution` or increase the window and stride as it typically allows for a much more accurate topic assignment.
